Git guide

northwestwitch opened the pull request:
Wrote a short guide to Git

northwestwitch commented:
@ingkebil @henrikstranneheim @robinandeer @moonso , I've expanded the git howto page. Wasn't unable to assign the pull request so I mention you guys! :)

Chiara

henrikstranneheim commented on git/README.md:
```diff
@@ -1,30 +1,230 @@
+
+
+## 1. Purpose
+This document offers an overview of themost important commands of Github and it is intended to be a quick start guide to work with Github repositories.
```
themost

henrikstranneheim commented on git/README.md:
```diff
@@ -1,30 +1,230 @@
+Your workspace is the folder that holds the actual working files while the "index" is the space is a staging area where you add your changes before committing them to the local repository or "head". To send the changes to your remote repository you use the "push" command.
+
+## 6. Keep your fork up to date.
+Before committing and pushing changes to your remote repository you should chack that you are working on an updated version of the repository. To update your fork follow these instructions:
```
chack->check

ingkebil replied:
This is the first time you mention a fork. What is a fork and how do create one?

henrikstranneheim commented on git/README.md:
```diff
@@ -1,30 +1,230 @@
+## 6. Keep your fork up to date.
+Before committing and pushing changes to your remote repository you should chack that you are working on an updated version of the repository. To update your fork follow these instructions:
+
+From command line CD to the main folder containing your project and type:
```
From the command line move to the main

henrikstranneheim commented on git/README.md:
```diff
@@ -1,30 +1,230 @@
 
-> you can also write `git checkout -b [my-topic]`.
+## 7. Create a new branch/feature for your work.
+To create a new branch, from command line CD to the main folder containing your project and type:
```
From the command line move to the main

northwestwitch commented:
Hi!
About the commands in bold and not, I did it this way to put more emphasis
on the most important ones.. But I realize now that this thing is clear
only in my mind! So I agree, I'll make them all bold!

Thanks!

Chiara


On Wed, Aug 30, 2017 at 2:04 PM, Henrik Stranneheim <
notifications@github.com> wrote:

> *@henrikstranneheim* commented on this pull request.
>
> Nice!
> Some commands are in "bold" and some are not. It would look better with
> consistency
> ------------------------------
>
> In git/README.md
> <https://github.com/Clinical-Genomics/development/pull/18#discussion_r136044291>
> :
>
> >
> -## Setting up this guide
> +
> +
> +## 1. Purpose
> +This document offers an overview of themost important commands of Github and it is intended to be a quick start guide to work with Github repositories.
>
> themost
> ------------------------------
>
> In git/README.md
> <https://github.com/Clinical-Genomics/development/pull/18#discussion_r136044810>
> :
>
> > +</pre>
> +
> +Where username@host:/path/to/repository is the url which can be obtained from the web page of your local fork, by clicking on the "clone or download" green button.
> +The git clone command then creates a folder with the project and pulls the latest data from your fork in it.
> +
> +
> +## 5. Git data transport structure and commands.
> +The following image shows the structure of the git commands and spaces:
> +
> +<p align="center">
> +  <img src="http://i.stack.imgur.com/MgaV9.png">
> +</p>
> +Your workspace is the folder that holds the actual working files while the "index" is the space is a staging area where you add your changes before committing them to the local repository or "head". To send the changes to your remote repository you use the "push" command.
> +
> +## 6. Keep your fork up to date.
> +Before committing and pushing changes to your remote repository you should chack that you are working on an updated version of the repository. To update your fork follow these instructions:
>
> chack->check
> ------------------------------
>
> In git/README.md
> <https://github.com/Clinical-Genomics/development/pull/18#discussion_r136044941>
> :
>
> > +Where username@host:/path/to/repository is the url which can be obtained from the web page of your local fork, by clicking on the "clone or download" green button.
> +The git clone command then creates a folder with the project and pulls the latest data from your fork in it.
> +
> +
> +## 5. Git data transport structure and commands.
> +The following image shows the structure of the git commands and spaces:
> +
> +<p align="center">
> +  <img src="http://i.stack.imgur.com/MgaV9.png">
> +</p>
> +Your workspace is the folder that holds the actual working files while the "index" is the space is a staging area where you add your changes before committing them to the local repository or "head". To send the changes to your remote repository you use the "push" command.
> +
> +## 6. Keep your fork up to date.
> +Before committing and pushing changes to your remote repository you should chack that you are working on an updated version of the repository. To update your fork follow these instructions:
> +
> +From command line CD to the main folder containing your project and type:
>
> From the command line move to the main
> ------------------------------
>
> In git/README.md
> <https://github.com/Clinical-Genomics/development/pull/18#discussion_r136045179>
> :
>
> >  ```
>
> -> you can also write `git checkout -b [my-topic]`.
> +## 7. Create a new branch/feature for your work.
> +To create a new branch, from command line CD to the main folder containing your project and type:
>
> From the command line move to the main
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Clinical-Genomics/development/pull/18#pullrequestreview-59527940>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AayssmtjTZgT18aP3KKsI_PCzl1qDJdXks5sdU-3gaJpZM4PHAIb>
> .
>

henrikstranneheim commented:
👍

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+
+Where username@host:/path/to/repository is the url which can be obtained from the web page of your local fork, by clicking on the "clone or download" green button.
+The git clone command then creates a folder with the project and pulls the latest data from your fork in it.
+
```
You could add an example here by giving the clone command to clone this repo

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+  <img src="http://i.stack.imgur.com/MgaV9.png">
+</p>
+Your workspace is the folder that holds the actual working files while the "index" is the space is a staging area where you add your changes before committing them to the local repository or "head". To send the changes to your remote repository you use the "push" command.
+
```
It would be nice to have a link that explains all the terms in the image.

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+git pull upstream develop (develop it's the version you start from for your work!)
+</b>
+</pre>
+
```
Could you split this up in setting you're fork to able to receive updates from the remote and how to update the fork? Now it's both in one and it is confusing.

ingkebil replied:
bump

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+
+
+You don't have to call the remote "upstream" but you can chose any name for it. To visualize all remotes you are working with and their respective shortnames use the command:
+
```
Maybe add a section that explains the common terms used to identify repos in Git: remote, master, upstream.

ingkebil replied:
...and origin :)

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+<pre>
+<b>git remote -v</b>
+</pre>
+   
```
This is really cool to know

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+<pre>
+<b>git checkout branch_to_use_for_your_work</b>
+</pre>
+
```
There is also a shorthand command called `git checkout -b branch_to_use_for_your_work`

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+</pre>
+
+## 8. Make changes and commit them to your local repository.
+Before committing any changes to your local repository, and finally to the remote repository, make sure you were working on a feature created from an updated version of the develop branch (see point 6: Keep your fork up to date).
```
There is prolly a way to reference that point directly?

ingkebil commented on git/README.md:
````diff
@@ -1,32 +1,229 @@
+```bash
+git diff source_branch target_branch
+```
+
````
Good to mention Git diff! One can also just call Git diff without any arguments to see the diff's of the current unstsged files with the HEAD.

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+</pre>
+
+The you can start adding your changes to the index with the command:
+
```
Adding a file to the index is called staging.

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+<pre>
+<b>git commit -a</b>
+</pre>
+
```
This should be used with care. I've involuntary added temporary files and lock files too many times like that that I've stopped using it.
At the very least one should `git status` and `git diff` before doing a commit -a.

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+<pre>
+<b>git push origin (or branch_you_want_to_push_to)</b>
+</pre>
+
```
If your remote called origin is configured and you're committing to mastee , you can leave that one out the word origin.

ingkebil commented on git/README.md:
```diff
@@ -1,32 +1,229 @@
+
+## 10. Create a pull request to the parent repository.
+You create a "pull request" when you want to propose the introduction of your changes to the repository where your project was forked from. Pull requests are created from the webpage of your remote repository. By default pull requests are directed to the default branch of the parent repository, so make sure to use properly the drop-down menu on the webpage to pull changes to the develop branch an ...
+
```
Good! Don't forget to link the github help page

northwestwitch commented:
I've hopefully addressed all your comments, which were very good by the
way!

Thanks guys, and if you still see something weird just tell me! :)

Chiara

On Wed, Aug 30, 2017 at 9:36 PM, Kenny Billiau <notifications@github.com>
wrote:

> *@ingkebil* commented on this pull request.
> ------------------------------
>
> In git/README.md
> <https://github.com/Clinical-Genomics/development/pull/18#discussion_r136168551>
> :
>
> > +## 6. Keep your fork up to date.
> +Before committing and pushing changes to your remote repository you should chack that you are working on an updated version of the repository. To update your fork follow these instructions:
> +
> +From command line CD to the main folder containing your project and type:
> +
> +<pre>
> +<b>git remote add upstream path/to/repo/you/forked/from
> +git fetch upstream
> +git pull upstream master
> +git pull upstream develop (develop it's the version you start from for your work!)
> +</b>
> +</pre>
> +
> +
> +You don't have to call the remote "upstream" but you can chose any name for it. To visualize all remotes you are working with and their respective shortnames use the command:
> +
>
> ...and origin :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Clinical-Genomics/development/pull/18#discussion_r136168551>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AayssiZQiERbPg1EoK9-7_dcJzl0k-Odks5sdbnOgaJpZM4PHAIb>
> .
>

henrikstranneheim commented on git/README.md:
```diff
@@ -189,49 +210,58 @@ To commit these files to the local repository type:
 
 <pre>
-<b>git commit -a</b>
+<b>git commit -a</b>*
```
I think you should remove the asterix from cmd call as it is not part of the actual command.
Maybe you could add NOTE: under the cmd line or some other visual clue that this command has a heads up

henrikstranneheim commented on git/README.md:
```diff
@@ -189,49 +210,58 @@ To commit these files to the local repository type:
+<b>git push origin (or branch_you_want_to_push_to)</b>**
 </pre>
 
+\** If your remote, called origin, is configured and you're committing to master, you can omit the word "origin" and just type "git push".
```
Same here for asterix

henrikstranneheim commented:
Looks good! For me you can go ahead and merge!

ingkebil commented on git/README.md:
```diff
@@ -1,40 +1,267 @@
+To create a new feature instead, type:
+
+<pre>
+<b>git flow init
```
Please mention that `git flow` is not part of the base package and is only available through `brew`. e.g. this will not work on rasta.

ingkebil commented on git/README.md:
```diff
@@ -1,40 +1,267 @@
+
+<pre>
+<b>brew install git
+brew install git-flow</b>
```
@Clinical-Genomics/owners is anyone using git-flow as a package?
I'm only familiar git `github flow` way of working, which is a way of working and not a pacakge. Is it worth exploring this package?

robinandeer replied:
I think we used it for scout a while ago but decided github flow was a better fit for us

I mentioned in a comment in am system that we should explain how to install git-flow before getting into examples 🙂

ingkebil replied:
aha cool.Well, if no-one is using git flow, we shouldn't mention it here. We can link to how github flow works instead.

ingkebil commented on git/README.md:
```diff
@@ -76,7 +76,8 @@ Create a directory with the name of the new repository and enter it from command
 ### 4.2 Clone a repository from an existing project.
-After forking a repository from the internet (how to fork a repository) you can create a local copy of it from command line, by moving in the parent folder of your new repository, then typing:
+When you are contributing to an existing repository, the first thing you should do is forking this repository to your space. A fork is basically a copy of the original project, on which you can experiment as much as you want without changing the original project.
+After forking a repository from the internet ([how to fork a repository](https://guides.github.com/activities/forking/)) you can create a local copy of it from command line, by moving in the parent folder of your new repository, then typing:
```
neat!

northwestwitch commented:
I was actually using it, but it's fine, I didn't have the time to grow
attached to it anyway ;)
I'll remove it!

On Thu, Aug 31, 2017 at 9:37 PM, Kenny Billiau <notifications@github.com>
wrote:

> *@ingkebil* commented on this pull request.
> ------------------------------
>
> In git/README.md
> <https://github.com/Clinical-Genomics/development/pull/18#discussion_r136428996>
> :
>
> >
> -## GitHub
> +Here are instructions on how to [install Git under other Linux distributions.](https://git-scm.com/download/linux)
> +
> +### 2.2 Download and install Git for OSX.
> +
> +The latest version of Git for Mac can be downloaded [here](https://git-scm.com/download/mac), use this binary installer to get an up-to date version of he software.
> +
> +If you are installing on a Mac with OS X 10.9 you can install Git from command line with Homebrew (get Homebrew [here](https://brew.sh/)) by typing:
> +
> +<pre>
> +<b>brew install git
> +brew install git-flow</b>
>
> aha cool.Well, if no-one is using git flow, we shouldn't mention it here.
> We can link to how github flow works instead.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Clinical-Genomics/development/pull/18#discussion_r136428996>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AaysspmEFyZACQIWG7TfUP0AGvb_Aqydks5sdwtfgaJpZM4PHAIb>
> .
>

northwestwitch commented:
Hi guys, could you review it again so we can merge it? Thanks!

henrikstranneheim commented:
I am going to go ahead and merge this as I think it not useful if these PR lay to long idle. Anyone that wants update can make another PR in the future when they have time.
